src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts

```diff
@@ -0,0 +1,333 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { TreeSitterTokenizationRegistry } from 'vs/editor/common/languages';
+import type { Parser } from '@vscode/tree-sitter-wasm';
+import { AppResourcePath, FileAccess, nodeModulesPath } from 'vs/base/common/network';
+import { ITreeSitterParserService } from 'vs/editor/common/services/treeSitterParserService';
+import { IModelService } from 'vs/editor/common/services/model';
+import { Disposable, DisposableMap, IDisposable } from 'vs/base/common/lifecycle';
+import { ITextModel } from 'vs/editor/common/model';
+import { IFileService } from 'vs/platform/files/common/files';
+import { IModelContentChangedEvent } from 'vs/editor/common/textModelEvents';
+import { ITelemetryService } from 'vs/platform/telemetry/common/telemetry';
+import { ILogService } from 'vs/platform/log/common/log';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { setTimeout0 } from 'vs/base/common/platform';
+import { importAMDNodeModule } from 'vs/amdX';
+
+const EDITOR_EXPERIMENTAL_PREFER_TREESITTER = 'editor.experimental.preferTreeSitter';
+const moduleLocationTreeSitter: AppResourcePath = `${nodeModulesPath}/@vscode/tree-sitter-wasm/wasm`;
+const moduleLocationTreeSitterWasm: AppResourcePath = `${moduleLocationTreeSitter}/tree-sitter.wasm`;
+
+class TextModelTreeSitter extends Disposable {
+	private _treeSitterTree: TreeSitterTree | undefined;
+	private _contentChangedListener: IDisposable | undefined;
+
+	constructor(readonly model: ITextModel,
+		private readonly _treeSitterParser: TreeSitterParser,
+		private readonly _treeSitterImporter: TreeSitterImporter,
+	) {
+		super();
+		this._register(this.model.onDidChangeLanguage(e => this._onDidChangeLanguage(e)));
+		this._onDidChangeLanguage({ newLanguage: this.model.getLanguageId() });
+	}
+
+	private _registerModelListeners() {
+		this._contentChangedListener = this.model.onDidChangeContent(e => this._onDidChangeContent(e));
+		this._onDidChangeContent();
+	}
+
+	private _unregisterModelListeners() {
+		this._contentChangedListener?.dispose();
+	}
+
+	private async _onDidChangeLanguage(e: { newLanguage: string }) {
+		if (this._treeSitterTree) {
+			this._treeSitterTree.dispose();
+		}
+		const language = await this._treeSitterParser.getLanguage(e.newLanguage);
+		if (!language || this.model.isDisposed()) {
+			this._unregisterModelListeners();
+			return;
+		}
+
+		const Parser = await this._treeSitterImporter.getParserClass();
+		this._treeSitterTree = new TreeSitterTree(new Parser(), language);
```

Comment:
If the language changes twice within a short amount of time, there could be race conditions here.
Especially if it takes longer to load the first language than the second one! (then `this._treeSitterTree` first would be the second language, and then the first language)